duckduckgo/BrowserServicesKit
# Sync Engine with support for syncing bookmarks

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/0/1204606874917765/f
iOS PR: https://github.com/duckduckgo/iOS/pull/1724
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1203
What kind of version bump will this require?: Major

**Optional**:

Tech Design URL:
* https://app.asana.com/0/481882893211075/1204311618174648/f
* https://app.asana.com/0/481882893211075/1204345998879965/f
* https://app.asana.com/0/481882893211075/1204306755229293/f
* https://app.asana.com/0/481882893211075/1204448565097516/f
* https://app.asana.com/0/0/1204501868567932/f

**Description**:
Changes to bookmarks:
* Updated Bookmarks model to v3 with added `isPendingDeletion` and `modifiedAt`
* Bookmarks are now soft-deleted at all times. When sync is off, a database cleaner runs on schedule
  and removes soft-deleted bookmarks.
* Client UIs are updated to not display soft-deleted bookmarks

Changes to Sync:
* Replaced `isAuthenticated` with `SyncAuthState` enum (inactive, setupNewAccount, addNewDevice, active)
* Added `SyncScheduling` and `DataProviding` public protocols. Made `Crypting` public too.
* Added `SyncMetadata` Core Data model, `SyncMetadataStore` class and `SyncFeatureEntity` class representing a syncable model.
* Sync is triggered from clients by calling scheduler APIs. Scheduler decides internally when to start sync.
* SyncQueue is an actor and it can only run a single sync operation at a time.

New module – SyncDataProviders:
* For now only contains `BookmarkProvider`, it’s an implementation of `DataProviding` for handling bookmarks.
  The logic is fully reused in iOS and macOS apps.



**Recommended reading order**:
1. BookmarkEntity.swift, BookmarkUtils.swift, BookmarksDatabaseCleaner.swift
  (other changes under Bookmarks models are related to iOS app UI)
3. DDGSyncing.swift, DDGSync.swift, ProductionDependencies.swift
4. SyncScheduler.swift, SyncQueue.swift
5. BookmarksProvider.swift
6. The rest.

**Steps to test this PR**:
To speed up testing sync you may want to set `SyncScheduler.appLifecycleEventsDebounceInterval` to 1.

1. Run client apps from develop and import some bookmarks.
1. Switch to `dominik/sync-engine` branches on both clients and run the app again. Make sure that the app didn’t crash at startup due to failed migration.
1. Enable sync on one of the devices. Then enable sync on another device by adding it to the first device’s account.
1. Make sure that devices’ data was merged, i.e. added device contains both local and remote bookmarks and vice-versa.
1. Make changes to data and verify that they are synced according to the design.



**OS Testing**:


—
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
